src/transport/raw/TCP.cpp

```diff
@@ -349,18 +347,16 @@ CHIP_ERROR TCPBase::ProcessSingleMessageFromBufferHead(const PeerAddress & peerA
 }
 
 CHIP_ERROR TCPBase::ProcessReceivedBuffer(Inet::TCPEndPoint * endPoint, const PeerAddress & peerAddress,
-                                          System::PacketBuffer * buffer)
+                                          System::PacketBufferHandle buffer)
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
 
-    while (buffer != nullptr)
+    while (!buffer.IsNull())
     {
         // when a buffer is empty, it can be released back to the app
         if (buffer->DataLength() == 0)
         {
-            System::PacketBuffer * old = buffer;
-            buffer                     = old->DetachTail();
-            System::PacketBuffer::Free(old);
+            buffer.Adopt(buffer->DetachTail());
```

Comment:
We might want, in a followup, to add a method to `PacketBufferHandle` that encapsulates this "drop the head" pattern.

Reply:
Yes, there are a whole bunch of DetachTail and AddToEnd that I intend to cover later. Most of them are on the sending path, so I left it alone for this PR.